### PR TITLE
F/kv parser robustness improvements, drop empty keys, better test coverage

### LINKED ISSUE #1072: Minor problems in kv-parser()
Hi,

I wrote a couple of edge-case unit tests for `kv-parser()`, some of them have produced questionable results.

- value-separator alone creates an empty key-value pair:
```C
static void
test_kv_scanner_stray_words_are_ignored(void)
{
  kv_scanner_input(kv_scanner, "lorem ipsum foo=bar");
  assert_next_kv_is("foo", "bar");
  assert_no_more_tokens();
// ...
  kv_scanner_input(kv_scanner, "lorem ipsum = foo=bar");
  assert_next_kv_is("", ""); // is it OK?
  assert_next_kv_is("foo", "bar");
  assert_no_more_tokens();
// ...
}
```

---------------

- a comma separated value without a space behaves differently:
```C
  kv_scanner_input(kv_scanner, "key1=value1,key2=value2,key3=value3");
  assert_next_kv_is("key1", "value1,key2=value2,key3=value3");
  assert_no_more_tokens();
```

---------------

- tabulator as a whitespace character is not working properly:
```C
  kv_scanner_input(kv_scanner, "key1=value1    key2=value2,     key3=value3");
  // key=value1 <tab> key2=value2, <space> <tab> key3=value3
  assert_next_kv_is("key1", "value1    key2=value2");
  assert_next_kv_is("key3", "value3");
  assert_no_more_tokens();
```

### PR DESCRIPTION
**Pull request by bkil-syslogng**
Fixes #1072

### CODE REVIEW

**Comment by bkil-syslogng** [comment]
Rebased to master. Data race fixes have been cherry picked on top. A leak fix is still expected by tomorrow, which will also be cherry picked on top, so reviewing the previous parts is still valid.

File: modules/kvformat/tests/test_kv_scanner.c [hunk 1]
```diff
@@ -20,196 +20,302 @@
...
+  kv_scanner_input(scanner, input);
+  gboolean expect_more = TRUE;
+  while (fn(scanner, args, &expect_more))
+    {
+    }
+  if (expect_more)
+    _assert_no_more_tokens(scanner);
 }
 
+#define _INVOKE_VA(args, code) \
```
Comment (kvch) [inline modules/kvformat/tests/test_kv_scanner.c]: `_INVOKE_VA(args, code, input)`

**Comment by kvch** [comment]
After the small mistake is fixed, this can be merged. :+1:

**Comment by bkil-syslogng** [comment]
Fixed comments. At first sight, the mentioned memory leak doesn't apply to this version of kv-parser, so this PR seems final. I'll proceed with investigation on a different track tomorrow, though.

**Comment by bkil-syslogng** [comment]
Another `kv-parser` race fixed. The `GString *formatted_key` scratch buffer was also shared. I tried to condition the fixes on whether threading is enabled or not to not cause a speed regression.

File: modules/kvformat/kv-parser.c [hunk 2]
```diff
@@ -112,6 +112,28 @@ kv_parser_free(LogPipe *s)
...
+
+  kv_parser_free(&self->super);
+  return ok;
+}
+
+static gboolean
+kv_parser_process_init(LogParser *s, LogMessage **pmsg, const LogPathOptions *path_options, const gchar *input, gsize input_len)
+{
+  if (s->super.cfg && s->super.cfg->threaded)
+    s->process = kv_parser_process_threaded;
```
Comment (bazsi) [inline modules/kvformat/kv-parser.c]: there are other ways to initiate threading (the threaded flag on an input driver), so this is not really safe.

And I would try to avoid any conditionals in driver code if at all possible.

Can't we avoid malloc/free by using a local variable when cloning? But you know what, only a small fraction of the input will be parsed by kv-parser(), so for now let's stick to the clone() stuff, always. We'll see if there are bottlenecks in the area. If there is, I would probably use a pool of KVScanner instances, one dedicated to all threads.

So I think the conditions on threaded should just be reverted.

**Comment by bazsi** [comment]
With the one comment on the conditionals, this could go in.
:+1:

**Comment by bkil-syslogng** [comment]
Thank you for your time. I've now removed the check and retained the threaded version only.

While the simple `kv_scanner_clone()` solution looked sane of first glance, unfortunately it would introduce a memory leak which would have been caught by https://github.com/balabit/syslog-ng/pull/1080
I've adjusted it to use `log_pipe_clone()` and `log_pipe_unref()` instead.

Can we now merge this with these two +1's?

[The previous fix proposals used local variables (and sprinkled conditioning) instead. I've opted for this cloning solution now because it is not much slower, while being much easier to understand and more difficult to mess up in the future.]

Backup saved:
* https://github.com/balabit/syslog-ng/compare/master...bkil-syslogng:t/kv-parser-robustness-bk3?expand=1
